File: unlz2k.cpp
#include "unlz2k.hpp"
#include <cstring>
#include <intrin.h>
#include <iostream>
#include <istream>

#define MAX_CHUNK_SIZE 0x40000

// Globals that are used by other functions

uint8_t compressedFile[MAX_CHUNK_SIZE];
size_t tmpSrcOffs;
size_t tmpSrcSize;
size_t tmpDestSize;
uint32_t bitstream;
uint8_t lastByteRead;
uint8_t previousBitAlign;
uint16_t chunksWithCurrentSetupLeft;
uint32_t readOffset;
int32_t literalsToCopy;
uint8_t tmpChunk[8192];
uint8_t smallByteDict[20];
uint8_t largeByteDict[510];
uint16_t smallWordDict[256];
uint16_t parallelDict0[1024];
uint16_t parallelDict1[1024];
uint16_t largeWordDict[4096];
uint32_t lz2k = 0x4C5A324B; // "LZ2K"

enum ENDIAN { LITTLE, BIG };

constexpr ENDIAN getEndianOrder() {
  return ((0xFFFFFFFF & 1) == 0x00000001) ? LITTLE : BIG;
}

#define _ENDIANNESS_ getEndianOrder()

size_t unlz2k_chunk(std::ifstream &, std::ofstream &, size_t, size_t);
void loadIntoBitstream(uint8_t);
uint32_t decodeBitstream();
uint32_t decodeBitstreamForLiterals();
void fillSmallDicts(uint8_t, uint8_t, uint8_t);
void fillLargeDicts();
void fillWordsUsingBytes(uint16_t, uint8_t *, uint8_t, uint16_t *);
void readAndDecrypt(size_t, uint8_t *);
void writeToFile(std::ofstream &, uint8_t *, size_t);

// Checks if program stores data natively as big or little endian
bool isNativeLE() {
  uint16_t num = 1;
  return *(char *)&num == 1;
}

// Read 32 bit unsigned integer from file
uint32_t readUint32(std::ifstream &src, ENDIAN endianness) {
  uint32_t data;
  src.read((char *)&data, 4);
  if (_ENDIANNESS_ != endianness) {
    return _byteswap_ulong(data);
  } else {
    return data;
  }
}

// Read 16 bit unsigned integer from file
uint16_t readUint16(std::ifstream &src, ENDIAN endianness) {
  uint16_t data;
  src.read((char *)&data, 2);
  if (_ENDIANNESS_ != endianness) {
    return _byteswap_ushort(data);
  } else {
    return data;
  }
}

// This version does not check integrity. It relies on the rest of the
// program!
size_t unlz2k(std::ifstream &src, std::ofstream &dest) {
  src.seekg(0, std::ios::end);
  size_t fileEnd = src.tellg();
  src.seekg(0);
  while (src.tellg() < fileEnd) {
    uint32_t header = readUint32(src, ENDIAN::BIG);
    if (header != lz2k) {
      std::cerr << "Not valid LZ2K file or chunk at pos "
                << (size_t)src.tellg() - 4 << "\n";
      return 1;
    }
    uint32_t unpacked = readUint32(src, ENDIAN::LITTLE);
    uint32_t packed = readUint32(src, ENDIAN::LITTLE);
    unlz2k_chunk(src, dest, packed, unpacked);
  }
  return 0;
}

size_t unlz2k(std::ifstream &src, std::ofstream &dest, size_t srcSize,
              size_t destSize) {
  size_t bytesWritten = 0;
  while (bytesWritten < destSize) {
    uint32_t header = readUint32(src, ENDIAN::BIG);
    if (header != lz2k) {
      std::cerr << "Not valid LZ2K file or chunk at pos "
                << (size_t)src.tellg() - 4 << "\n";
      return 1;
      return 1;
    }
    uint32_t unpacked = readUint32(src, ENDIAN::LITTLE);
    uint32_t packed = readUint32(src, ENDIAN::LITTLE);
    bytesWritten += unlz2k_chunk(src, dest, packed, unpacked);
  }
  if (src.tellg() != srcSize) {
    printf("Expected to read 0x%08x bytes, but wrote 0x%08x", srcSize,
           (size_t)src.tellg());
    return 1;
  }
  if (bytesWritten != destSize) {
    printf("Expected to write 0x%08x bytes, but wrote 0x%08x", destSize,
           bytesWritten);
    return 1;
  }
  return 0;
}

size_t unlz2k_chunk(std::ifstream &src, std::ofstream &dest, size_t srcSize,
                    size_t destSize) {
  if (!destSize) {
    return 0;
  }
  // Read file into memory location
  src.read((char *)compressedFile, srcSize);
  tmpSrcOffs = 0;
  tmpSrcSize = srcSize;
  bitstream = 0;
  lastByteRead = 0;
  previousBitAlign = 0;
  chunksWithCurrentSetupLeft = 0;
  readOffset = 0;
  literalsToCopy = 0;
  size_t bytesLeft = destSize;
  size_t bytesWritten = 0;
  loadIntoBitstream(32);
  while (bytesLeft > 0) {
    size_t chunkSize = (bytesLeft < 8192 ? bytesLeft : 8192);
    readAndDecrypt(chunkSize, tmpChunk);
    writeToFile(dest, tmpChunk, chunkSize);
    bytesWritten += chunkSize;
    bytesLeft -= chunkSize;
  }
  return bytesWritten;
}

void loadIntoBitstream(uint8_t bits) {
  bitstream <<= bits;
  if (bits > previousBitAlign) {
    do {
      bits -= previousBitAlign;
      bitstream |= lastByteRead << bits;
      if (tmpSrcOffs == tmpSrcSize) {
        lastByteRead = 0;
      } else {
        lastByteRead = compressedFile[tmpSrcOffs++];
      }
      previousBitAlign = 8;
    } while (bits > previousBitAlign);
  }
  previousBitAlign -= bits;
  bitstream |= lastByteRead >> previousBitAlign;
}

void readAndDecrypt(size_t chunkSize, uint8_t *out) {
  uint32_t outputOffs = 0;
  --literalsToCopy;
  if (literalsToCopy >= 0) {
    do {
      out[outputOffs++] = out[readOffset++];
      readOffset &= 0x1FFF;
      if (outputOffs == chunkSize) {
        return;
      }
      literalsToCopy--;
    } while (literalsToCopy >= 0);
  }
  while (outputOffs < chunkSize) {
    uint32_t tmpVal = decodeBitstream();
    if (tmpVal <= 255) {
      out[outputOffs++] = tmpVal;
      if (outputOffs == chunkSize)
        return;
    } else {
      uint32_t toCopy = decodeBitstreamForLiterals();
      readOffset = (outputOffs - toCopy - 1) & 0x1FFF;
      toCopy = tmpVal - 254;
      literalsToCopy = toCopy;
      while (literalsToCopy >= 0) {
        out[outputOffs++] = out[readOffset++];
        readOffset &= 0x1FFF;
        if (outputOffs == chunkSize)
          return;
        literalsToCopy--;
      }
    }
  }
  if (outputOffs > chunkSize) {
    std::cerr << "Error: went farther than given length\n";
    return;
  }
}

void writeToFile(std::ofstream &outFile, uint8_t *data, size_t length) {
  outFile.write((const char *)data, length);
}

uint32_t decodeBitstream() {
  if (!chunksWithCurrentSetupLeft) {
    chunksWithCurrentSetupLeft = bitstream >> 16;
    loadIntoBitstream(16);
    fillSmallDicts(19, 5, 3);
    fillLargeDicts();
    fillSmallDicts(14, 4, -1);
  }
  chunksWithCurrentSetupLeft--;
  uint16_t tmpVal = largeWordDict[bitstream >> 20];
  if (tmpVal >= 510) {
    uint32_t mask = 0x80000;
    do {
      if (!(bitstream & mask)) {
        tmpVal = parallelDict0[tmpVal];
      } else {
        tmpVal = parallelDict1[tmpVal];
      }
      mask >>= 1;
    } while (tmpVal >= 510);
  }
  uint8_t bits = largeByteDict[tmpVal];
  loadIntoBitstream(bits);
  return tmpVal;
}

uint32_t decodeBitstreamForLiterals() {
  uint8_t tmpOffs = bitstream >> 24;
  uint16_t tmpVal = smallWordDict[tmpOffs];
  if (tmpVal >= 14) {
    uint32_t mask = 0x800000;
    do {
      if (!(bitstream & mask)) {
        tmpVal = parallelDict0[tmpVal];
      } else {
        tmpVal = parallelDict1[tmpVal];
      }
      mask >>= 1;
    } while (tmpVal >= 14);
  }
  uint8_t bits = smallByteDict[tmpVal];
  loadIntoBitstream(bits);
  if (!tmpVal)
    return 0;
  if (tmpVal == 1)
    return 2;
  tmpVal--;
  uint32_t tmpBitstream = bitstream >> (32 - tmpVal);
  loadIntoBitstream(tmpVal);
  return tmpBitstream + (1 << tmpVal);
}

void fillSmallDicts(uint8_t length, uint8_t bits, uint8_t specialInd) {
  uint32_t tmpVal = bitstream >> (32 - bits);
  loadIntoBitstream(bits);
  if (!tmpVal) {
    tmpVal = bitstream >> (32 - bits);
    loadIntoBitstream(bits);
    if (length > 0) {
      memset(smallByteDict, 0, length);
    }
    for (int i = 0; i < 256; ++i) {
      smallWordDict[i] = tmpVal;
    }
    return;
  }
  uint32_t tmpVal2 = 0;
  while (tmpVal2 < tmpVal) {
    uint8_t tmpByte = bitstream >> 29;
    uint8_t bits = 3;
    if (tmpByte == 7) {
      uint32_t mask = 0x10000000;
      if (!(bitstream & mask)) {
        bits = 4;
      } else {
        uint8_t counter = 0;
        do {
          mask >>= 1;
          counter++;
        } while (bitstream & mask);
        bits = counter + 4;
        tmpByte += counter;
      }
    }
    loadIntoBitstream(bits);
    smallByteDict[tmpVal2++] = tmpByte;
    if (tmpVal2 == specialInd) {
      size_t specialLen = bitstream >> 30;
      loadIntoBitstream(2);
      if (specialLen) {
        memset(smallByteDict + tmpVal2, 0, specialLen);
        tmpVal2 += specialLen;
      }
    }
  }
  if (tmpVal2 < length) {
    memset(smallByteDict + tmpVal2, 0, length - tmpVal2);
  }
  fillWordsUsingBytes(length, smallByteDict, 8, smallWordDict);
  return;
}

void fillLargeDicts() {
  int16_t tmpVal = bitstream >> 23;
  loadIntoBitstream(9);
  if (!tmpVal) {
    tmpVal = bitstream >> 23;
    loadIntoBitstream(9);
    memset(largeByteDict, 0, 510);
    for (int i = 0; i < 4096; ++i) {
      largeWordDict[i] = tmpVal;
    }
    return;
  }
  uint16_t bytes = 0;
  if (tmpVal < 0) {
    // Does this ever happen?
    memset(largeByteDict, 0, 510);
    fillWordsUsingBytes(510, largeByteDict, 12, largeWordDict);
    return;
  }
  while (bytes < tmpVal) {
    uint8_t tmpLen = bitstream >> 24;
    uint16_t tmpVal2 = smallWordDict[tmpLen];
    if (tmpVal2 >= 19) {
      uint32_t mask = 0x800000;
      do {
        if (!(bitstream & mask)) {
          tmpVal2 = parallelDict0[tmpVal2];
        } else {
          tmpVal2 = parallelDict1[tmpVal2];
        }
        mask >>= 1;
      } while (tmpVal2 >= 19);
    }
    uint8_t bits = smallByteDict[tmpVal2];
    loadIntoBitstream(bits);
    if (tmpVal2 > 2) {
      tmpVal2 -= 2;
      largeByteDict[bytes++] = tmpVal2;
    } else {
      if (!tmpVal2) {
        tmpLen = 1;
      } else if (tmpVal2 == 1) {
        tmpVal2 = bitstream >> 28;
        loadIntoBitstream(4);
        tmpLen = tmpVal2 + 3;
      } else {
        tmpVal2 = bitstream >> 23;
        loadIntoBitstream(9);
        tmpLen = tmpVal2 + 20;
      }
      if (tmpLen) {
        memset(largeByteDict + bytes, 0, tmpLen);
        bytes += tmpLen;
      }
    }
  }
  if (bytes < 510) {
    memset(largeByteDict + bytes, 0, 510 - bytes);
  }
  fillWordsUsingBytes(510, largeByteDict, 12, largeWordDict);
  return;
}

void fillWordsUsingBytes(uint16_t bytesLen, uint8_t *bytes, uint8_t pivot,
                         uint16_t *words) {
  uint16_t srcDict[17] = {0};
  uint16_t destDict[18];
  destDict[1] = 0;
  for (int i = 0; i < bytesLen; ++i) {
    uint8_t tmp = bytes[i];
    srcDict[tmp]++;
  }
  uint8_t shift = 14, ind = 1;
  uint16_t low, high;
  while (ind <= 16) {
    low = srcDict[ind] << (shift + 1);
    high = srcDict[ind + 1] << shift;
    low += destDict[ind];
    ind += 4;
    high += low;
    destDict[ind - 3] = low;
    destDict[ind - 2] = high;
    low = srcDict[ind - 2] << (shift - 1);
    low += high;
    high = srcDict[ind - 1] << (shift - 2);
    high += low;
    destDict[ind - 1] = low;
    destDict[ind] = high;
    shift -= 4;
  }
  if (destDict[17]) {
    std::cerr << "Bad table\n";
  }
  shift = pivot - 1;
  uint8_t tmpVal = 16 - pivot;
  uint8_t tmpValCopy = tmpVal;
  for (int i = 1; i <= pivot; ++i) {
    destDict[i] >>= tmpValCopy;
    srcDict[i] = 1 << shift--;
  }
  tmpValCopy--;
  for (int i = pivot + 1; i <= 16; ++i) {
    srcDict[i] = 1 << tmpValCopy--;
  }
  uint16_t comp1 = destDict[pivot + 1];
  comp1 >>= 16 - pivot;
  if (comp1) {
    uint16_t comp2 = 1 << pivot;
    if (comp1 != comp2) {
      for (int i = comp1; i < comp2; ++i) {
        words[i] = 0;
      }
    }
  }
  if (bytesLen <= 0)
    return;
  shift = 15 - pivot;
  uint16_t mask = 1 << shift;
  uint16_t bytesLenCopy = bytesLen;
  for (int i = 0; i < bytesLen; ++i) {
    uint8_t tmpByte = bytes[i];
    if (tmpByte) {
      uint16_t destVal = destDict[tmpByte];
      uint16_t srcVal = srcDict[tmpByte] + destVal;
      if (tmpByte > pivot) {
        uint16_t *dictPtr = words;
        uint16_t tmpOffs = destVal >> tmpVal;
        uint8_t newLen = tmpByte - pivot;
        while (newLen) {
          if (!dictPtr[tmpOffs]) {
            parallelDict0[bytesLenCopy] = 0;
            parallelDict1[bytesLenCopy] = 0;
            dictPtr[tmpOffs] = bytesLenCopy++;
          }
          tmpOffs = dictPtr[tmpOffs];
          if (!(destVal & mask)) {
            dictPtr = parallelDict0;
          } else {
            dictPtr = parallelDict1;
          }
          destVal += destVal;
          newLen--;
        }
        dictPtr[tmpOffs] = i;
      } else if (destVal < srcVal) {
        for (int j = destVal; j < srcVal; ++j) {
          words[j] = i;
        }
      }
      destDict[tmpByte] = srcVal;
    }
  }
}

File: unlz2k.py
""" Python implementation of the extraction algorithm used by TT Games """

import os, sys, time

########
# DATA #
########

tmp_src = None # Used by load_into_bitstream
tmp_src_size = None # Used by load_into_bitstream
bitstream = None
last_byte_read = None
previous_bit_align = None
chunks_with_current_setup = None
read_offset = None
literals_to_copy = None
tmp_chunk = None
byte_dict0 = None # 510 byte dictionary
byte_dict1 = None # 20 byte dictionary before tempChunk
word_dict0 = None # 256 word dictionary
word_dict1 = None # 1024 word dictionary
word_dict2 = None # 1024 word dictionary
word_dict3 = None # 4096 word dictionary

#############
# FUNCTIONS #
#############

def memset(src, src_offset, value, length):
    for i in range(length):
        src[src_offset + i] = value

def load_into_bitstream(bits):
    global bitstream
    global previous_bit_align
    global last_byte_read
    global tmp_src
    global tmp_src_size
    bitstream <<= bits
    last_op_diff = previous_bit_align
    data = last_byte_read
    if bits > last_op_diff:
        while bits > last_op_diff:
            bits -= last_op_diff
            data <<= bits
            bitstream |= data
            if tmp_src_size == 0:
                data = 0
            else:
                tmp_src_size -= 1
                data = int.from_bytes(tmp_src.read(1), byteorder="little")
            last_op_diff = 8
        last_byte_read = data
    last_op_diff -= bits
    bits = last_op_diff
    data >>= bits
    previous_bit_align = last_op_diff
    bitstream |= data
    # 32 bit align
    bitstream &= 0xFFFFFFFF

def process_dicts(val0, dict0, val1, dict1):
    global word_dict1
    global word_dict2
    tmp_src_dict = dict.fromkeys(range(17), 0)
    tmp_dest_dict = dict.fromkeys(range(18), 0)
    tmp_dest_dict[1] = 0
    for i in range(val0):
        tmp = dict0[i]
        tmp_src_dict[tmp] += 1
    shift = 14
    ind = 1
    while ind <= 16:
        low = tmp_src_dict[ind]
        high = tmp_src_dict[ind + 1]
        low <<= shift + 1
        high <<= shift
        low += tmp_dest_dict[ind]
        ind += 4
        high += low
        high &= 0xFFFF
        tmp_dest_dict[ind - 3] = low
        low = tmp_src_dict[ind - 2]
        low <<= shift - 1
        tmp_dest_dict[ind - 2] = high
        low += high
        low &= 0xFFFF
        high = tmp_src_dict[ind - 1]
        high <<= shift - 2
        tmp_dest_dict[ind - 1] = low
        high += low
        high &= 0xFFFF
        tmp_dest_dict[ind] = high
        shift -= 4
    if tmp_dest_dict[17] != 0:
        print('Bad table')
        exit(1)
    shift = val1 - 1
    tmp_val = 16 - val1
    var_5c = tmp_val
    for i in range(1, val1 + 1):
        tmp_dest_dict[i] >>= tmp_val
        tmp_src_dict[i] = 1 << shift
        shift -= 1
    tmp_val = 16 - (val1 + 1)
    for i in range(val1 + 1, 17):
        tmp_src_dict[i] = 1 << tmp_val
        tmp_val -= 1
    comp1 = tmp_dest_dict[val1 + 1]
    comp1 >>= 16 - val1
    if comp1 != 0:
        comp2 = 1 << val1
        if comp1 != comp2:
            for i in range(comp1, comp2):
                dict1[i] = 0
    if val0 <= 0:
        return
    shift = 15 - val1
    mask = 1 << shift
    tmp_val0 = val0
    for i in range(0, val0):
        ind = dict0[i]
        if (ind != 0):
            tmp_dest_val = tmp_dest_dict[ind]
            tmp_src_val = tmp_src_dict[ind]
            tmp_src_val += tmp_dest_val
            if ind > val1:
                tmp_dict = dict1
                tmp_offs = tmp_dest_val >> var_5c
                new_len = ind - val1
                if new_len != 0:
                    while new_len != 0:
                        if tmp_dict[tmp_offs] == 0:
                            word_dict1[tmp_val0] = 0
                            word_dict2[tmp_val0] = 0
                            tmp_dict[tmp_offs] = tmp_val0
                            tmp_val0 += 1
                        tmp_offs = tmp_dict[tmp_offs]
                        if tmp_dest_val & mask == 0:
                            tmp_dict = word_dict1
                        else:
                            tmp_dict = word_dict2
                        tmp_dest_val += tmp_dest_val
                        new_len -= 1
                tmp_dict[tmp_offs] = i
            elif tmp_dest_val < tmp_src_val:
                for j in range(tmp_dest_val, tmp_src_val):
                    dict1[j] = i
            tmp_dest_dict[ind] = tmp_src_val

def setup_byte_and_word_dicts(length, bits, special_ind):
    global bitstream
    global byte_dict1
    global word_dict0
    tmp_val = bitstream >> (32 - bits) # bits is never 0
    load_into_bitstream(bits)
    if tmp_val != 0:
        tmp_val2 = 0
        if tmp_val > 0:
            while tmp_val2 < tmp_val:
                tmp_bitstream = bitstream >> 29
                bits = 3
                if tmp_bitstream == 7:
                    mask = 0x10000000
                    if bitstream & mask == 0:
                         bits = 4
                    else:
                        counter = 0
                        while bitstream & mask != 0:
                            mask >>= 1
                            counter += 1
                        bits = counter + 4
                        tmp_bitstream += counter
                load_into_bitstream(bits)
                byte_dict1[tmp_val2] = tmp_bitstream
                tmp_val2 += 1
                if tmp_val2 == special_ind:
                    special_len = bitstream >> 30
                    load_into_bitstream(2)
                    if special_len >= 1:
                        memset(byte_dict1, tmp_val2, 0, special_len)
                        tmp_val2 += special_len
        if tmp_val2 < length:
            memset(byte_dict1, tmp_val2, 0, length - tmp_val2)
        process_dicts(length, byte_dict1, 8, word_dict0)
        return
    tmp_val = bitstream >> (32 - bits)
    load_into_bitstream(bits)
    if length > 0:
        memset(byte_dict1, 0, 0, length)
    for i in range(256):
        word_dict0[i] = tmp_val

def setup_byte_dict0():
    global bitstream
    global byte_dict0
    global byte_dict1
    global word_dict0
    global word_dict1
    global word_dict2
    global word_dict3
    tmp_val = bitstream >> 23
    load_into_bitstream(9)
    if tmp_val == 0:
        tmp_val2 = bitstream >> 23
        load_into_bitstream(9)
        memset(byte_dict0, 0, 0, 510)
        for i in range(4096):
            word_dict3[i] = tmp_val2
        return
    num_bytes = 0
    if tmp_val > 0: # always?
        while num_bytes < tmp_val:
            tmp_len = bitstream >> 24
            tmp_val2 = word_dict0[tmp_len]
            if tmp_val2 >= 19:
                mask = 0x800000
                while tmp_val2 >= 19:
                    if mask & bitstream == 0:
                        tmp_val2 = word_dict1[tmp_val2]
                    else:
                        tmp_val2 = word_dict2[tmp_val2]
                    mask >>= 1
            bits = byte_dict1[tmp_val2]
            load_into_bitstream(bits)
            if tmp_val2 > 2:
                tmp_val2 -= 2
                byte_dict0[num_bytes] = tmp_val2
                num_bytes += 1
            else:
                if tmp_val2 == 0:
                    tmp_len = 1
                else:
                    if tmp_val2 != 1:
                        tmp_val2 = bitstream >> 23
                        load_into_bitstream(9)
                        tmp_len = tmp_val2 + 20
                    else:
                        tmp_val2 = bitstream >> 28
                        load_into_bitstream(4)
                        tmp_len = tmp_val2 + 3
                if tmp_len > 0:
                    memset(byte_dict0, num_bytes, 0, tmp_len)
                    num_bytes += tmp_len
        if num_bytes < 510:
            memset(byte_dict0, num_bytes, 0, 510 - num_bytes)
        process_dicts(510, byte_dict0, 12, word_dict3)
        return
    memset(byte_dict0, 0, 0, 510)
    process_dicts(510, byte_dict0, 12, word_dict3)
    return

def decode_bitstream():
    global chunks_with_current_setup
    global bitstream
    global word_dict1
    global word_dict2
    global word_dict3
    if chunks_with_current_setup == 0:
        chunks_with_current_setup = bitstream >> 16
        load_into_bitstream(16)
        setup_byte_and_word_dicts(19, 5, 3)
        setup_byte_dict0()
        setup_byte_and_word_dicts(14, 4, -1)
    chunks_with_current_setup -= 1
    tmp_val = word_dict3[bitstream >> 20]
    if tmp_val >= 510:
        mask = 0x80000
        while (tmp_val >= 510):
            if bitstream & mask == 0:
                tmp_val = word_dict1[tmp_val]
            else:
                tmp_val = word_dict2[tmp_val]
            mask >>= 1
    bits = byte_dict0[tmp_val]
    load_into_bitstream(bits)
    return tmp_val

def decode_bitstream_for_literals():
    global bitstream
    global byte_dict1
    global word_dict0
    global word_dict1
    global word_dict2
    tmp_offs = bitstream >> 24
    tmp_val = word_dict0[tmp_offs]
    if tmp_val >= 14:
        mask = 0x800000
        while tmp_val >= 14:
            if bitstream & mask == 0:
                tmp_val = word_dict1[tmp_val]
            else:
                tmp_val = word_dict2[tmp_val]
            mask >>= 1
    bits = byte_dict1[tmp_val]
    load_into_bitstream(bits)
    if tmp_val == 0:
        return 0
    if tmp_val == 1:
        return 2
    tmp_val -= 1
    tmp_bitstream = bitstream >> (32 - tmp_val)
    load_into_bitstream(tmp_val)
    return tmp_bitstream + (1 << tmp_val)

def read_and_decrypt(length, dest):
    global read_offset
    global literals_to_copy
    output_offs = 0
    tmp_read_offs = read_offset
    literals_to_copy -= 1
    tmp_to_copy = literals_to_copy
    if tmp_to_copy >= 0:
        while tmp_to_copy >= 0:
            dest[output_offs] = dest[tmp_read_offs]
            output_offs += 1
            tmp_read_offs += 1
            tmp_read_offs &= 0x1FFF
            if output_offs == length:
                literals_to_copy = tmp_to_copy
                read_offset = tmp_read_offs
                return
            tmp_to_copy -= 1
        literals_to_copy = tmp_to_copy
        read_offset = tmp_read_offs
    while output_offs < length:
        tmp_val = decode_bitstream()
        if tmp_val <= 255:
            dest[output_offs] = tmp_val
            output_offs += 1
            if output_offs == length:
                return
        else:
            tmp_to_copy = decode_bitstream_for_literals()
            tmp_read_offs = output_offs - tmp_to_copy - 1
            tmp_to_copy = tmp_val - 254
            tmp_read_offs &= 0x1FFF
            read_offset = tmp_read_offs
            literals_to_copy = tmp_to_copy
            while tmp_to_copy >= 0:
                dest[output_offs] = dest[tmp_read_offs]
                output_offs += 1
                tmp_read_offs += 1
                tmp_read_offs &= 0x1FFF
                read_offset = tmp_read_offs
                if output_offs == length:
                    return
                tmp_to_copy -= 1
                literals_to_copy = tmp_to_copy
    if output_offs > length:
        print('Error: went farther than given length')

def copy_to_file(file, src, length):
    """ Writes bytearray data of given length to file """
    file.write(src[:length])

def unlz2k_chunk(src, dest, src_size, dest_size):
    if (dest_size == 0):
        return 0
    # Shared variables
    global tmp_src
    tmp_src = src
    global tmp_src_size
    tmp_src_size = src_size
    global bitstream
    bitstream = 0
    global last_byte_read
    last_byte_read = 0
    global previous_bit_align
    previous_bit_align = 0
    global chunks_with_current_setup
    chunks_with_current_setup = 0
    global read_offset
    read_offset = 0
    global literals_to_copy
    literals_to_copy = 0
    global tmp_chunk
    tmp_chunk = bytearray(8192)
    # Shared dicts
    global byte_dict0
    byte_dict0 = dict.fromkeys(range(510), 0)
    global byte_dict1
    byte_dict1 = dict.fromkeys(range(20), 0)
    global word_dict0
    word_dict0 = dict.fromkeys(range(256), 0)
    global word_dict1
    word_dict1 = dict.fromkeys(range(1024), 0)
    global word_dict2
    word_dict2 = dict.fromkeys(range(1024), 0)
    global word_dict3
    word_dict3 = dict.fromkeys(range(4096), 0)
    # Local variables
    bytes_left = dest_size
    bytes_written = 0
    
    load_into_bitstream(32)
    while (bytes_left != 0):
        chunk_size = bytes_left if bytes_left < 8192 else 8192
        read_and_decrypt(chunk_size, tmp_chunk)
        copy_to_file(dest, tmp_chunk, chunk_size)
        bytes_written += chunk_size
        bytes_left -= chunk_size
    return bytes_written

def unlz2k(src, dest, src_size, dest_size):
    bytes_written = 0
    while bytes_written < dest_size:
        lz2k = src.read(4)
        if (lz2k != b'LZ2K'):
            print('Not valid LZ2K file or chunk at {:<6X}'.format(src.tell()))
            exit(1)
        unpacked = int.from_bytes(src.read(4), byteorder="little")
        packed = int.from_bytes(src.read(4), byteorder="little")
        bytes_written += unlz2k_chunk(src, dest, packed, unpacked)
    bytes_read = src.tell() - original_src
    if bytes_read != src_size:
        print('Size mismatch, got {:<8X} but read {:<8X}'.format(src_size, bytes_read))
        exit(1)
    if bytes_written != dest_size:
        print('Size mismatch, got {:<8X} but wrote {:<8X}'.format(dest_size, bytes_written))
        exit(1)
    print('Done.')

# Does not check bytes read or written
def unlz2k(src, dest):
    src.seek(0, 2)
    end = src.tell()
    src.seek(0)
    while src.tell() < end:
        lz2k = src.read(4)
        if (lz2k != b'LZ2K'):
            print('Not valid LZ2K file or chunk at {:<6X}'.format(src.tell()))
            exit(1)
        unpacked = int.from_bytes(src.read(4), byteorder="little")
        packed = int.from_bytes(src.read(4), byteorder="little")
        unlz2k_chunk(src, dest, packed, unpacked)
    print('Done.')

################
# TEST PROGRAM #
################

""" if len(sys.argv) < 2:
    print('Please specify a filename.')
    exit(1)
filename = sys.argv[1]
src = open(filename, 'rb')
dest = open(filename + '.dec', 'wb')
unlz2k(src, dest)
src.close()
dest.close()
"""
